File: src/solar/solar-common/util/math/LeapUtil.ts
import BaseUtil from "../BaseUtil";
import GroupUtil from "../unit/GroupUtil";
import SelectUtil from "../unit/SelectUtil";
import EffectUtil from "@/EffectUtil";
import DamageType from "@/DamageType";
import DamageUtil from "@/DamageUtil";
import RectUtil from "@/RectUtil";

/**
 * 冲锋
 */
/** @noSelf **/
export default class LeapUtil {

    /**
     *
     * @param damage_src_handle
     * @param angle 传入角度
     * @param range
     * @param time
     * @param damage
     * @param dmgType
     * @param effectPath  // 有特效则 只移动特效 比如发射投射物冲锋
     * @param startX
     * @param startY
     * @param onHitUnit //当冲击到敌方单位时 （可以在这里做击飞等操作）
     */
    static leap(damage_src_handle: unit, angle: number, range: number, time: number, damage: number = 0, dmgType: DamageType = DamageType.s_物理,
                effectPath?: string, startX: number = GetUnitX(damage_src_handle), startY: number = GetUnitY(damage_src_handle), onHitUnit?: (enemy) => void) {
        let execCount = time * 25
        let dis = range / execCount;
        let effectHandle: effect = null
        if (effectPath && effectPath.length > 0) {
            effectHandle = AddSpecialEffect(effectPath, startX, startY)
            EffectUtil.setEffectFacing(effectHandle, angle)
        }
        let damagedUnitGroup: group = null;
        if (damage > 0) {
            damagedUnitGroup = GroupUtil.groupObjectPool.borrowObject();
        }
        BaseUtil.onTimer(0.04, (c) => {
            let x = 0;
            let y = 0;
            if (effectHandle) {
                x = dis * Math.cos(angle) + EXGetEffectX(effectHandle);
                y = dis * Math.sin(angle) + EXGetEffectY(effectHandle)
                if (RectUtil.isInPlayableArea(x, y)) {
                    EXSetEffectXY(effectHandle, x, y);
                }
            } else {
                x = dis * Math.cos(angle) + GetUnitX(damage_src_handle);
                y = dis * Math.sin(angle) + GetUnitY(damage_src_handle)
                if (RectUtil.isInPlayableArea(x, y) && LeapUtil.isPathAble(x, y)) {
                    SetUnitPosition(damage_src_handle, x, y);
                }
            }
            if (damage > 0) {
                SelectUtil.forEnemyUnitsInRange(damage_src_handle, 128, unit => {
                    //说明已经伤害过了
                    if (IsUnitInGroup(unit, damagedUnitGroup)) {
                        return
                    }
                    GroupAddUnit(damagedUnitGroup, unit);
                    DamageUtil.damage(damage_src_handle, unit, damage, dmgType)
                    onHitUnit?.(unit)
                }, x, y);
            }
            //DestroyTimer
            execCount = execCount - 1;
            if (execCount < 1) {
                if (damagedUnitGroup) {
                    GroupUtil.groupObjectPool.returnObject(damagedUnitGroup);
                }
                if (effectHandle) {
                    DestroyEffect(effectHandle)
                }
                return false;
            }
            return true;
        })
    }

    /**
     * 指定点是否可以冲锋
     * @param x
     * @param y
     */
    static isPathAble(x: number, y: number) {
        return !IsTerrainPathable(x, y, PATHING_TYPE_WALKABILITY);
    }

    /**
     * 无限冲锋 (通常为boss的技能使用 把一个玩家打飞到地图边界才停止了 以感受boss的强悍)
     * @param unit
     * @param angle
     * @param speed 每秒距离
     * @param panCamera 平移镜头
     * @param onDone 完成时的回调
     */
    static infiniteLeap(unit: unit, angle: number, speed: number = 1000, panCamera: boolean = false, onDone?: () => void) {
        UnitAddAbility(unit, 'Amrf');
        SetUnitFlyHeight(unit, 256, 1000);
        EXPauseUnit(unit, true)
        //每帧距离 speed为每秒距离
        let dis = speed * 0.25;
        if (panCamera && GetOwningPlayer(unit) != GetLocalPlayer()) {
            panCamera = false;
        }
        BaseUtil.onTimer(0.04, (c) => {
            let x = dis * Math.cos(angle) + GetUnitX(unit);
            let y = dis * Math.sin(angle) + GetUnitY(unit)
            if (RectUtil.isInPlayableArea(x, y)) {
                SetUnitPosition(unit, x, y);
                if (panCamera) {
                    PanCameraToTimed(x, y, 0)
                }
            } else {
                EXPauseUnit(unit, false)
                SetUnitFlyHeight(unit, 0, 1000);
                UnitRemoveAbility(unit, 'Amrf');
                onDone?.()
                return false;
            }
            return true;
        })
    }


}

File: scripts/solar/solar-common/util/math/LeapUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 1,["7"] = 1,["8"] = 2,["9"] = 2,["10"] = 3,["11"] = 3,["12"] = 4,["13"] = 4,["14"] = 5,["15"] = 5,["16"] = 6,["17"] = 6,["18"] = 7,["19"] = 7,["20"] = 13,["21"] = 13,["22"] = 13,["24"] = 13,["25"] = 28,["26"] = 28,["27"] = 28,["29"] = 28,["30"] = 28,["32"] = 29,["33"] = 29,["35"] = 29,["36"] = 29,["38"] = 30,["39"] = 31,["40"] = 32,["41"] = 33,["42"] = 34,["43"] = 35,["45"] = 37,["46"] = 38,["47"] = 39,["49"] = 41,["50"] = 41,["51"] = 41,["52"] = 42,["53"] = 43,["54"] = 44,["55"] = 45,["56"] = 46,["57"] = 47,["58"] = 48,["61"] = 51,["62"] = 52,["63"] = 53,["64"] = 54,["67"] = 57,["68"] = 58,["69"] = 58,["70"] = 58,["71"] = 58,["72"] = 60,["75"] = 63,["76"] = 64,["77"] = 65,["78"] = 65,["80"] = 58,["81"] = 58,["82"] = 58,["83"] = 58,["85"] = 69,["86"] = 70,["87"] = 71,["88"] = 72,["90"] = 74,["91"] = 75,["93"] = 77,["95"] = 79,["96"] = 41,["97"] = 41,["98"] = 28,["99"] = 88,["100"] = 89,["101"] = 88,["102"] = 100,["103"] = 100,["104"] = 100,["106"] = 100,["107"] = 100,["109"] = 101,["110"] = 102,["111"] = 103,["112"] = 105,["113"] = 106,["114"] = 107,["116"] = 109,["117"] = 109,["118"] = 109,["119"] = 110,["120"] = 111,["121"] = 112,["122"] = 113,["123"] = 114,["124"] = 115,["127"] = 118,["128"] = 119,["129"] = 120,["130"] = 121,["131"] = 121,["133"] = 122,["135"] = 124,["136"] = 109,["137"] = 109,["138"] = 100});
local ____exports = {}
local ____BaseUtil = require("solar.solar-common.util.BaseUtil")
local BaseUtil = ____BaseUtil.default
local ____GroupUtil = require("solar.solar-common.util.unit.GroupUtil")
local GroupUtil = ____GroupUtil.default
local ____SelectUtil = require("solar.solar-common.util.unit.SelectUtil")
local SelectUtil = ____SelectUtil.default
local ____EffectUtil = require("solar.solar-common.util.game.EffectUtil")
local EffectUtil = ____EffectUtil.default
local ____DamageType = require("solar.solar-common.constant.DamageType")
local DamageType = ____DamageType.default
local ____DamageUtil = require("solar.solar-common.util.system.DamageUtil")
local DamageUtil = ____DamageUtil.default
local ____RectUtil = require("solar.solar-common.util.game.RectUtil")
local RectUtil = ____RectUtil.default
____exports.default = __TS__Class()
local LeapUtil = ____exports.default
LeapUtil.name = "LeapUtil"
function LeapUtil.prototype.____constructor(self)
end
function LeapUtil.leap(damage_src_handle, angle, range, time, damage, dmgType, effectPath, startX, startY, onHitUnit)
    if damage == nil then
        damage = 0
    end
    if dmgType == nil then
        dmgType = DamageType["s_物理"]
    end
    if startX == nil then
        startX = GetUnitX(damage_src_handle)
    end
    if startY == nil then
        startY = GetUnitY(damage_src_handle)
    end
    local execCount = time * 25
    local dis = range / execCount
    local effectHandle = nil
    if effectPath and #effectPath > 0 then
        effectHandle = AddSpecialEffect(effectPath, startX, startY)
        EffectUtil:setEffectFacing(effectHandle, angle)
    end
    local damagedUnitGroup = nil
    if damage > 0 then
        damagedUnitGroup = GroupUtil.groupObjectPool:borrowObject()
    end
    BaseUtil.onTimer(
        0.04,
        function(____, c)
            local x = 0
            local y = 0
            if effectHandle then
                x = dis * math.cos(angle) + EXGetEffectX(effectHandle)
                y = dis * math.sin(angle) + EXGetEffectY(effectHandle)
                if RectUtil.isInPlayableArea(x, y) then
                    EXSetEffectXY(effectHandle, x, y)
                end
            else
                x = dis * math.cos(angle) + GetUnitX(damage_src_handle)
                y = dis * math.sin(angle) + GetUnitY(damage_src_handle)
                if RectUtil.isInPlayableArea(x, y) and ____exports.default.isPathAble(x, y) then
                    SetUnitPosition(damage_src_handle, x, y)
                end
            end
            if damage > 0 then
                SelectUtil.forEnemyUnitsInRange(
                    damage_src_handle,
                    128,
                    function(____, unit)
                        if IsUnitInGroup(unit, damagedUnitGroup) then
                            return
                        end
                        GroupAddUnit(damagedUnitGroup, unit)
                        DamageUtil:damage(damage_src_handle, unit, damage, dmgType)
                        if onHitUnit ~= nil then
                            onHitUnit(nil, unit)
                        end
                    end,
                    x,
                    y
                )
            end
            execCount = execCount - 1
            if execCount < 1 then
                if damagedUnitGroup then
                    GroupUtil.groupObjectPool:returnObject(damagedUnitGroup)
                end
                if effectHandle then
                    DestroyEffect(effectHandle)
                end
                return false
            end
            return true
        end
    )
end
function LeapUtil.isPathAble(x, y)
    return not IsTerrainPathable(x, y, PATHING_TYPE_WALKABILITY)
end
function LeapUtil.infiniteLeap(unit, angle, speed, panCamera, onDone)
    if speed == nil then
        speed = 1000
    end
    if panCamera == nil then
        panCamera = false
    end
    UnitAddAbility(unit, "Amrf")
    SetUnitFlyHeight(unit, 256, 1000)
    EXPauseUnit(unit, true)
    local dis = speed * 0.25
    if panCamera and GetOwningPlayer(unit) ~= GetLocalPlayer() then
        panCamera = false
    end
    BaseUtil.onTimer(
        0.04,
        function(____, c)
            local x = dis * math.cos(angle) + GetUnitX(unit)
            local y = dis * math.sin(angle) + GetUnitY(unit)
            if RectUtil.isInPlayableArea(x, y) then
                SetUnitPosition(unit, x, y)
                if panCamera then
                    PanCameraToTimed(x, y, 0)
                end
            else
                EXPauseUnit(unit, false)
                SetUnitFlyHeight(unit, 0, 1000)
                UnitRemoveAbility(unit, "Amrf")
                if onDone ~= nil then
                    onDone(nil)
                end
                return false
            end
            return true
        end
    )
end
return ____exports
